warehouse/query-core/src/main/java/datawave/query/jexl/nodes/QueryPropertyMarker.java

```diff
@@ -161,6 +165,55 @@ protected void setupSource(JexlNode source) {
         this.jjtAddChild(refExpNode1, 0);
     }
     
+    /**
+     * Determines if this source node is already marked with the provided marker class
+     *
+     * @param source
+     *            a JexlNode
+     * @param markerClass
+     *            an instance of a {@link QueryPropertyMarker}
+     * @return true if this node is already marked with the provided marker class
+     */
+    public static boolean isSourceMarked(JexlNode source, Class<? extends QueryPropertyMarker> markerClass) {
+        return QueryPropertyMarker.findInstance(source).isType(markerClass);
+    }
+    
+    /**
+     * Determine if one the source node's ancestors is already marked. For example, is this subtree already marked as delayed?
+     *
+     * @param node
+     *            a JexlNode
+     * @return true
+     */
+    public static boolean isAncestorMarked(JexlNode node, Class<? extends QueryPropertyMarker> markerClass) {
```

Comment:
This method could be really expensive depending on how many steps up the tree we have to go. Maybe note this in the comments.

Reply:
alternatively add a cut off depth to the search for the marker in the visitor? So we don't do the same work multiple times.

Reply:
Updated docs forthcoming